### PR TITLE
Fix Maximum Scan Line CRTC register behaviour

### PR DESCRIPTION
**Pull request by johnnovak**
# Description

This is embarrassing. Apparently, I suck at bit-twiddlery...

Well, third time's (?) the charm! Right? RIGHT???! 😅 

Ok, so the non-double-scanning path was wrong; this broke the end of the intro (spaceship part) and the final flyby scene of **Second Reality** (and probably many other things).

Thanks to @Ringdingcoder for reporting this! 🏅 

As far as I can tell, now this behaves the same way as the old code before I "cleaned it up" 😅 

## Related issues

https://github.com/dosbox-staging/dosbox-staging/pull/3356
https://github.com/dosbox-staging/dosbox-staging/issues/3268

# Manual testing

- Watched **Second Reality** in its entirety at 12k cycles — perfect (apart from the palette bug in the intro)
- **Legend of Kyrandia 1** intro scroller — no glitches
- **Catacomb 3D** — no vertical doubling in-game
- **Ravenloft: Stone Prophet** — works fine
- **Ravenloft: Strahd's Posession** — never been better
- **Threat** — zero issues
- **The Summoning** — ditto

# Checklist

_Please tick the items as you have addressed them. Don't remove items; leave the ones that are not applicable unchecked._

I have:

- [ ] followed the project's [contributing guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/master/CONTRIBUTING.md) and [code of conduct](https://github.com/dosbox-staging/dosbox-staging/blob/master/CODE_OF_CONDUCT.md).
- [x] performed a self-review of my code.
- [x] commented on the particularly hard-to-understand areas of my code.
- [ ] split my work into well-defined, bisectable commits, and I [named my commits well](https://github.com/dosbox-staging/dosbox-staging/blob/main/CONTRIBUTING.md#commit-messages).
- [x] applied the appropriate labels (bug, enhancement, refactoring, documentation, etc.)
- [ ] [checked](https://github.com/dosbox-staging/dosbox-staging/blob/main/scripts/compile_commits.sh) that all my commits can be built.
- [ ] confirmed that my code does not cause performance regressions (e.g., by running the Quake benchmark).
- [ ] added unit tests where applicable to prove the correctness of my code and to avoid future regressions.
- [ ] made corresponding changes to the documentation or the website according to the [documentation guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md).
- [ ] [locally verified](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md#previewing-documentation-changes-locally) my website or documentation changes.

### CODE REVIEW

File: src/hardware/vga_crtc.cpp [hunk 1]
```diff
@@ -205,24 +205,13 @@ void vga_write_p3d5(io_port_t, io_val_t value, io_width_t)
...
-			// been changed:
-			//   - LSB of `maximum_scan_line`
-			//   - `start_vertical_blanking_bit9`
-			//   - `is_scan_doubling_enabled`
-			// and all other bits are unchanged.
-			//
-			// This revised criteria was taken from DOSBox X. The
-			// old DOSBox criteria was "if ALL bits have been
-			// changed EXCEPT for `line_compare_bit9`"; I can't see
-			// how could that possibly work.
```
Comment (johnnovak) [inline src/hardware/vga_crtc.cpp:223]: Oh, that comment... Clearly, I was confused. If you replace "ALL" with "ANY" in my sentence, that's the answer 😏

**Comment by johnnovak** [comment]
@Ringdingcoder

Related: it would be nice if you could publish the final compiled release for your fixed version of Second Reality:
https://github.com/Ringdingcoder/SecondReality

It would be nice to be able to run it in DOSBox without the palette bug, but sure as hell I won't try to build it myself 😄 You could just upload it as a release artifact in your repo (if it's not already available somewhere).

**Review by weirddan455 — APPROVED** [review]
Man... this VGA code.  I approved one of your recent PRs which I guess was wrong in some way?  I just have very limited understanding of what this code is doing exactly.  You probably have a better understanding than anyone else on the team.  I think the rest of us are scared to touch these files (at least I am) :sweat_smile: 

Anyway, I don't see anything obviously wrong with this and you've done some good testing so hopefully everything is correct this time.

**Comment by Ringdingcoder** [comment]
> Related: it would be nice if you could publish the final compiled release for your fixed version of Second Reality: https://github.com/Ringdingcoder/SecondReality
> 
> It would be nice to be able to run it in DOSBox without the palette bug, but sure as hell I won't try to build it myself 😄 You could just upload it as a release artifact in your forked repo (if it's not already available somewhere).

I never actually fixed the palette bug. What I did was mostly clean up the mess that used to be hundreds of unused and interim files and a horrible manual build process of building some files multiple times with different switches #defined because of data table precalculation. I also created replacements for some of the simple binary-only build tools used in the tree. The obfuscation build process is still a horrible mess, though.

Additionally, the published code is older and buggier than the released binary version from 1993 and has additional glitches in the scroller part at the beginning, one of them coincidentally hitting at the same time as the palette bug. These and a few other minor cosmetic things I fixed by reverse engineering the original. What I did not fix are Esc handling to terminate the demo and starting at a specific part by passing an argument "1"-"5" on the command line. Both of them lead to music and demo playback becoming de-synchronized.

I noticed something in my first screenshot, though: There is a hex dump of a segment:offset address in the middle of the screen, which would indicate that the previous part crashes here, which is something I was not aware of so far. So the palette bug might actually be caused by a crash. I think I will have another look then.

I could of course upload a binary, but for now it would not be an improvement over the original.

**Comment by Ringdingcoder** [comment]
FWIW, I tested my build against this, and it started working again, as I would have expected.

**Comment by johnnovak** [comment]
> Additionally, the published code is older and buggier than the released binary version from 1993

Doesn't surprise me, 99% of demos were written by enthusiastic teenagers having access to their own hardware only. There wasn't much "testing" going into these productions. Just has to work on the compo machine 😎 

> Man... this VGA code. I approved one of your recent PRs which I guess was wrong in some way? I just have very limited understanding of what this code is doing exactly. You probably have a better understanding than anyone else on the team. I think the rest of us are scared to touch these files (at least I am) 😅

I cleaned up the original bit twiddling stuff that used various tricks with bitmask to detect if "all of a set of bits" or "any of a set of bits" have changed between the old and new value. My mistake was confusing the "all" condition with "any".

Well, you can't reason about low level emulation code like this, just review it for style or something 😄 I'm far from being a VGA expert, but I've done some lightweight VGA programming back in the day, hitting VGA registers directly, demoscene style. When it comes to low-level hardware emulation, there's no rhyme or reasons to it, just making sure the generated output in emulation is the same on hardware for the same input combos. Well, the original devs did that here, probably, and I screwd it a bit up 😄

**Comment by Ringdingcoder** [comment]
> Related: it would be nice if you could publish the final compiled release for your fixed version of Second Reality: https://github.com/Ringdingcoder/SecondReality
> 
> It would be nice to be able to run it in DOSBox without the palette bug, but sure as hell I won't try to build it myself 😄 You could just upload it as a release artifact in your forked repo (if it's not already available somewhere).

Not sure if you noticed, but I [did that](https://github.com/Ringdingcoder/SecondReality/releases/tag/v0.1rc1) a while ago (with the palette bug [fixed](https://github.com/Ringdingcoder/SecondReality/commit/c7a553d975f756533f56ea04e9814f66cda454af)).

**Comment by johnnovak** [comment]
> Not sure if you noticed, but I [did that](https://github.com/Ringdingcoder/SecondReality/releases/tag/v0.1rc1) a while ago (with the palette bug [fixed](https://github.com/Ringdingcoder/SecondReality/commit/c7a553d975f756533f56ea04e9814f66cda454af)).

I did not, so thanks for pinging me about it @Ringdingcoder.

**Pull request merged**